libwebauthn/src/transport/nfc/pcsc/mod.rs

```diff
@@ -0,0 +1,167 @@
+        match card {
+            Some(card) => {
+                debug!("Disconnect card");
+                card.disconnect(pcsc::Disposition::UnpowerCard)
+                    .map_err(PcscCard::map_disconnect_error)
+            }
+            None => Ok(()),
+        }
+    }
+}
+
+pub struct Channel {
+    card: Arc<Mutex<PcscCard>>,
+}
+
+unsafe impl Send for Channel {}
+
+impl fmt::Display for Info {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.name)
+    }
+}
+
+impl From<pcsc::Error> for Error {
+    fn from(input: pcsc::Error) -> Self {
+        trace!("{:?}", input);
+        let output = match input {
+            pcsc::Error::NoSmartcard => TransportError::ConnectionFailed,
+            _ => TransportError::InvalidFraming,
+        };
+
+        Error::Transport(output)
+    }
+}
+
+impl Info {
+    pub fn new(name: &CStr) -> Self {
+        Info {
+            name: CStr::into_c_string(name.into()),
+        }
+    }
+
+    pub fn channel(&self) -> Result<(NfcChannel<Context>, mpsc::Receiver<UxUpdate>), Error> {
+        let (send, recv) = mpsc::channel(1);
+        let context = pcsc::Context::establish(pcsc::Scope::User)?;
+        let chan = Channel::new(self, context)?;
+
+        let ctx = Context {};
+        let channel = NfcChannel::new(Box::new(chan), ctx, send);
+        Ok((channel, recv))
+    }
+}
+
+impl Channel {
+    pub fn new(info: &Info, context: pcsc::Context) -> Result<Self, Error> {
+        let card = context.connect(&info.name, pcsc::ShareMode::Shared, pcsc::Protocols::ANY)?;
+
+        let chan = Self {
+            card: Arc::new(Mutex::new(PcscCard::new(card))),
+        };
+
+        Ok(chan)
+    }
+}
+
+impl fmt::Display for Channel {
+    fn fmt(&self, f: &mut fmt::Formatter) -> std::fmt::Result {
+        let card = self.card.lock().unwrap();
+        let (names_len, atr_len) = card.status2_len().unwrap();
+        let mut names_buf = vec![0; names_len];
+        let mut atr_buf = vec![0; atr_len];
+        let status = card.status2(&mut names_buf, &mut atr_buf).unwrap();
+        write!(f, "{:?}", status.reader_names().collect::<Vec<_>>())
+    }
+}
+
+impl<Ctx> NfcBackend<Ctx> for Channel where Ctx: fmt::Debug + fmt::Display {}
+
+impl<Ctx> HandlerInCtx<Ctx> for Channel
+where
+    Ctx: fmt::Debug + fmt::Display,
+{
+    fn handle_in_ctx(
+        &mut self,
+        _ctx: Ctx,
+        command: &[u8],
+        response: &mut [u8],
+    ) -> apdu_core::Result {
+        trace!("TX: {:?}", command);
+
+        let rapdu = self
+            .card
+            .lock()
+            .unwrap()
+            .transmit(command, response)
+            .map_err(|e| HandleError::Nfc(Box::new(e)))?;
+
+        trace!("RX: {:?}", rapdu);
+        Ok(rapdu.len())
+    }
+}
+
+#[instrument]
+pub fn list_devices() -> Result<Vec<NfcDevice>, Error> {
+    let ctx = pcsc::Context::establish(pcsc::Scope::User).expect("PC/SC context");
```

Comment:
Have a look here about the context: https://pcsclite.apdu.fr/api/group__API.html#gaa1b8970169fd4883a6dc4a8f43f19b67